### DISCUSSION TITLE
Esp32 mqtt delay in publishing data

### DISCUSSION THREAD

**Discussion by parthesh-hub** [emqx/emqx#6939] [Q&A]
Hello ,

We are currently working on a project that uses IMU sensors along with ESP32 Node MCU. We expected the behavior that, whenever our sensor will send data (in our case every 100ms) it should get published on EMQX mqtt cloud broker without any delay. But there is a delay in data getting published. Because we see that sensor is sending 10 msgs per second but on the emqx server we see messages getting published at random intervals.. 

Because of this behavior, there is delay in further process (which in our case is subscribing the published data into our reactJS app in realtime).

Following attached are the snapshots depicting the issue.

1) Serial Monitor show data getting published every 100 ms.
Here 6479 6583 6693 (approx every 100ms)
![image](https://user-images.githubusercontent.com/55626780/152971671-42be0960-bf62-466b-ad18-f68b91584358.png)

2) In the EMQX mqtt broker the messages are getting published at random intervals and not exactly after every 100ms.
here 48:318   48:623   49:019  .. such random publishing rate
![image](https://user-images.githubusercontent.com/55626780/152971895-22b6b135-a194-4fd9-aa0c-fe66101dbb88.png)

Fortunately there is no data loss ... but because of this behavior there is unfortunate delay..

Can anyone please help to find the solution fir this :)

**Comment by zmstone** [comment]
Hi @parthesh-hub 
We'd like to collect some more information:
* which version is in use ?
* how many nodes do you have in your cluster ? 
* how many subscribers are there ?

I have some in-general reasoning:
It's QoS 0, the publisher (from both ends, client or broker), claiming messages being sent "successfully" actually means the message has been put to TCP send buffer.
And it's hard to know where exactly the hiccup is.
One way to troubleshoot is to enable debug logs everywhere, and try to find out where in the message flow the latency started its divergence.

**Reply by parthesh-hub** [reply]
Thank you for the reply!!

We are using following versions:
ESp32 version : 1.0.6
MQTT version : 3.1.1
broker : EMQX  (currently using trial version)
hosted on : aws 
Nodes : 1
Subscribers : 2 
QoS : 0

In the EMQx brokers debug log there is nothing.

We also taught there is issue with some kind of buffer.. but aren't able to debug.. 
I hope its not because we are sending messages at very fast rate(per 100ms), as it's the need of project.

**Reply by zmstone** [reply]
Hi @parthesh-hub
You can request EMQX Cloud team's assistance to help/troubleshoot changing logging level to debut.
we can try to investigate from there.

In the meantime, you can start EMQX in docker (or install one in a Linux VM) to test the latency.

e.g. `docker run --name emqx -d --network host -e EMQX_LOG__LEVEL=debug emqx/emqx:4.3.11`

**Reply by parthesh-hub** [reply]
We checked the TCP packet sent time in Wireshark. And yes it is the issue. because it accumulates 2/3/4  packets in a buffer and send them together. Thats why we are having the uncertain behavior at broker, which receives messaged at uneven rates even though we are sending them from esp32 at rate of 100ms.

Attached is the image..
![wireshark](https://user-images.githubusercontent.com/55626780/153132092-3dc29da3-a00a-4159-9f35-76b5ef89b03f.jpeg)
 
 You can see in the image each TCP packet is sent after 300/400ms and contains 3/4 messages together. 
 
 What we expect is sending TCP packet every 100ms with only message in it.  Also the acknowledgement is received in time sufficient time, so I don't think it is something related to acknowledgement. 
 
 
Is there anything we can do to sent the packets at once rather them accumulating in buffer. Because speed is most important in the context of project.

**Reply by zmstone** [reply]
Hi @parthesh-hub 
this is great insight.
could you confirm check if the publisher client has delays on wire ?
in the meantime, please request EQMX cloud team to turn on debug log so we can exclude EMQX internal buffering (before reaching TCP buffer).

**Reply by parthesh-hub** [reply]
Thanks for the reply @zmstone ...  Actually We didin't understand ur qn about "could you confirm check if the publisher client has delays on wire ?"  
We are using wifi and the ESP32 board is directly connected to port of Laptop. Also we tried using multiple brokers and the issue exists, so as per our understanding its not related to broker side. (We are using trial version of emqx for testing)

**Reply by zmstone** [reply]
Hi @parthesh-hub 
If I understand it correctly, what you observed was: messages (PSH in wireshark capture) sent from 35.156 (EMQX cloud) to 192.168 (subscriber, the laptop?) had unstable delay. 
So I was wondering if your publisher (the laptop) 192.168 is buffering messages towards 35.156 (EMQX cloud).

**Reply by zmstone** [reply]
> per our understanding its not related to broker side.

I am not sure how you came to this conclusion, I'm afraid I need more elaboration.

**Comment by joshemq** [comment]
Hi @parthesh-hub, were you able to resolve this issue? If so, would you be open to sharing more details about your project with me? I'm EMQ's marketing director in the US, and I'd appreciate the chance to interview you and maybe share your project with others if you'd be interested. Would it be okay to contact you directly?

**Comment by Zelmoghazy** [comment]
Hey @parthesh-hub I am working on a very similar problem, did you find any solutions ?
